File: src/Lessons/3-TimeComplexity/3.2-PermMissingElem.go
package main

import (
	"fmt"
	"sort"
)

func main() {

	fmt.Println(solution([]int{1, 2, 11, 14, 3, 4, 5, 6, 7, 8, 9, 10, 12}))
}

func solution(someArray []int) int {
	if len(someArray) == 0 {
		return 1
	}
	missingNumber := 0
	//sort our array
	sort.Ints(someArray)
	for index := 0; index < len(someArray); index++ {
		number := index + 1
		if someArray[index] != number {
			missingNumber = number
			break
		}
	}
	//if no missing number, the missing number is the next largest number
	if missingNumber == 0 {
		missingNumber = someArray[len(someArray)-1] + 1
	}
	return missingNumber
}

File: src/Lessons/3-TimeComplexity/3.2-PermMissingElem.js
//Perm Missing Elem
//Array A contains N integers of [1..(n+1)]
//find the missing integer
function solution(A) {
    let array = A.sort((a, b) => a - b);

    let missingInteger = 0;
    if (array.length < 1) return 1;
    console.log()
    for (let index = 0; index < array.length; index++) {
        const integer = index + 1;
        console.log(index)
        if (array[index] !== integer) {
            missingInteger = integer;
            break;
        }
    }

    if (missingInteger == 0) missingInteger = array[array.length - 1] + 1;

    return missingInteger;
}

console.log(solution([1, 2]))